### PR TITLE
Refactor promise compartment

### LINKED ISSUE #23167: Create a marker type that indicates that a compartment has been entered
https://github.com/servo/servo/issues/22982 is an example of crashes that can occur if we use certain JS APIs without entering a compartment. This can often be challenging to notice because the JS engine will enter compartments for us automatically when content code invokes DOM APIs (like when a script has `let promise = some_element.requestFullScreen();`). That means that when the the Element::RequestFullScreen method is invoked from JS, the promise can be created safely without entering a compartment. However, if any Rust code invokes this same method without passing through the JS engine first, we can crash. I propose we introduce the following to help:
```rust
pub struct AlreadyInCompartment(());
impl AlreadyInCompartment {
    pub fn assert(global: &GlobalScope) -> AlreadyInCompartment {
        unsafe {
            assert!(!GetCurrentRealmOrNull(global.cx()).is_null());
        }
        AlreadyInCompartment
    }
}

pub enum InCompartment<'a> {
    Already(&'a AlreadyInCompartment),
    Entered(&'a AutoCompartment),
}

impl<'a> InCompartment<'a> {
    pub fn in_compartment<'a>(token: &'a AlreadyInCompartment) -> InCompartment<'a> {
        InCompartment::Already(token)
    }

    pub fn entered<'a>(token: &'a AutoCompartment) -> InCompartment<'a> {
        InCompartment::Entered(token)
    }
}
```

Then we can start adding `&InCompartment` arguments to APIs like `Promise::new_with_current_compartment`, which will force callers to prove that they have entered a compartent and allow us to remove the unsafety from that API. This can propagate outwards, adding these arguments to more and more callers to make all of our bindings more safe. Eventually we can modify the DOM bindings code generation to create an `AlreadyInCompartment` value and pass it to the Rust method that gets called.

### PR DESCRIPTION
**Pull request by AZWN**
This PR adds a mechanism to verify that certain code is executed inside a ```JSAutoCompartment```, and applies this to the ```Promise::new_in_current_compartment``` constructor.

r? @jdm 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #23167 

- [x] These changes do not require tests because they do not change existing functionality.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23253)

### CODE REVIEW

**Comment by AZWN** [comment]
I have one open question/issue. The available ```Promise``` constructors now look like this:

```
impl Promise {
    pub fn new(global: &GlobalScope, comp: &InCompartment) -> Rc<Promise> {
        Promise::new_in_current_compartment(global, comp)
    }

    #[allow(unsafe_code)]
    pub fn new_in_current_compartment(global: &GlobalScope, _comp: &InCompartment) -> Rc<Promise> {
        // ...
    }
    // ...
```

To, me, having both ```Promise::new``` and ```Promise::new_in_current_compartment``` looks a bit redundant, since the first is just delegating to the second. Would it be better to remove ```Promise::new_in_current_compartment```?

**Comment by jdm** [comment]
Promise::new should not accept an InCompartment argument. Instead it should enter the compartment for `global` before calling new_in_current_compartment.

**Comment by AZWN** [comment]
Ok, so If I understand you well, it should use ```InCompartment::Entered``` like his:
impl Promise {
    pub fn new(global: &GlobalScope) -> Rc<Promise> {
        let comp = JSAutoCompartment::with_obj(&global.get_cx(), &global);
        Promise::new_in_current_compartment(global, &InCompartment::Entered(&comp))
    }

    #[allow(unsafe_code)]
    pub fn new_in_current_compartment(global: &GlobalScope, _comp: &InCompartment) -> Rc<Promise> {
        // ...
    }
    // ...
}
```?

**Comment by jdm** [comment]
That's the idea! I think we can pass `InCompartment` by value, rather than by reference, however.

**Comment by AZWN** [comment]
Ok, give me a moment, and I will implement it. 

I am not sure about the fact ```InCompartment``` should be passed by value though. The advantage of references in my opinion is that it is possible to reuse ```InCompartment``` values for multiple calls. For now, that is not that relevant yet, but if you are planning to extend this pattern to more use cases it might become an advantage.

**Comment by jdm** [comment]
If you add `#[derive(Copy, Clone)]` to the InCompartment struct definition, it should be possible to reuse them.

**Comment by jdm** [comment]
Fantastic! I can already see places where we can improve our safety guarantees based on these changes. Thanks for doing this work!
@bors-servo r+

**Comment by jdm** [comment]
```
error[E0308]: mismatched types
  --> components/script/body.rs:57:9
   |
57 |         &InCompartment::Already(&in_compartment_proof),
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         expected enum `compartments::InCompartment`, found reference
   |         help: consider removing the borrow: `InCompartment::Already(&in_compartment_proof)`
   |
   = note: expected type `compartments::InCompartment<'_>`
              found type `&compartments::InCompartment<'_>`

error[E0599]: no function or associated item named `with_obj` found for type `js::jsapi::JSAutoCompartment` in the current scope
  --> components/script/dom/promise.rs:84:39
   |
84 |         let comp = JSAutoCompartment::with_obj(&global.get_cx(), &global);
   |                                       ^^^^^^^^ function or associated item not found in `js::jsapi::JSAutoCompartment`

error: aborting due to 2 previous errors
```

File: components/script/dom/promise.rs [hunk 1]
```diff
@@ -80,12 +80,13 @@ impl Drop for Promise {
 }
 
 impl Promise {
-    pub fn new(global: &GlobalScope, comp: &InCompartment) -> Rc<Promise> {
-        Promise::new_in_current_compartment(global, comp)
+    pub fn new(global: &GlobalScope) -> Rc<Promise> {
+        let comp = AlreadyInCompartment::assert(&global);
```
Comment (jdm) [inline components/script/dom/promise.rs]: We want to always enter a compartment, not panic if we have not entered one. We need to use `let compartment = JSAutoCompartment::new(cx, global.reflector().get_jsobject().get());` here.

**Review by jdm — CHANGES_REQUESTED** [review]

**Pull request merged**